### PR TITLE
fix #19373 move all hairpins, when inserting measure into part

### LINKED ISSUE #19373: Inseriting measure(s) in a part not move hairpins on score
### Issue type

Other type of issue

### Bug description

MuseScore 4.1.1 not "move" the crescendo or diminuendo hairpin in the score (in other parts) if a measure is inserted into a part.

### Steps to reproduce

1. Insert a crescendo or diminuendo hairpin in the score
2. Open a part
3. Insert a measure (i.e. pressing _Insert_ in keaboard)
4. Return to the score

### Screenshots/Screen recordings

Insert a crescendo or diminuendo hairpin in the score
image

Open a part and insert a measure(s) into the part
image

Returning to the score only the hairpin of the edited part was moved in the score
image

### MuseScore Version

4.1.1

### Regression

No

### Operating system

Windows 11, Arch Linux

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sammik**
Resolves: #19373
Resolves: #19587
Resolves partialy (second issue) of #19476

~~undoInsertTime "acts on the linked scores as well", but only in score, there are all elements to be linked~~
inserting Measures always acts in `masterScore` and adds measures to all scores
inserting Frames acts in local score

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [x] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

**Comment by sammik** [comment]
this is somehow similar as #19271

question is, if `undoInsertTime` should also be always called from master score, or in other cases, it is good to call it from "local" score

@cbjeukendrup ?

**Comment by cbjeukendrup** [comment]
Yes, it would make sense to me to perform all operations relating to inserting/removing measures via the MasterScore, so inserting and deleting measures too. 
Same for the "remove trailing empty measures" command: #17350

**Comment by sammik** [comment]
> Yes, it would make sense to me to perform all operations relating to inserting/removing measures via the MasterScore, just like in #19271. 

It is not possible, as `insertMeasureOtions` have property `addToAllScores`, so in some situations, it is performed on just a particular excerpt score.

I can imagine something like
```
Score::insertMeasure(ElementType type, MeasureBase* beforeMeasure, const InsertMeasureOptions& options)
{
    if (options.addToAllScores && !score().isMaster()) {
        ...
        masterScore().insertMeasures(insertMeasure(type, masterBeforeMeasure, options);
        ...
        return ...;
    }
    ...
}
```
Not sure, if it is good idea.
@cbjeukendrup

**Comment by sammik** [comment]
> Perhaps we should split the method into two methods: one for the elementary operation of adding measures to one `Score` instance only, and one for the score editing operation of adding measures globally. The elementary local operation should be in the Score class; the score editing operation should be in the MasterScore class. The global operation should probably call the local operation for all scores from `scoresList()`. You might need to add another method to `Score` that is a wrapper around the global method, for example to facilitate selecting the newly inserted measures in the current score.
> 
> What do you think about that idea?

It seems to me a bit overcomplicated, and somehow I dont see many benefits over [my current implementation](https://github.com/musescore/MuseScore/pull/19405/files), but sure, I can try.

**Comment by sammik** [comment]
A little bit a different solution.

I found some problems with Your propose - there are bits of code, which would make sense to be called just once, in both cases - after we do some job on just local score / multiple scores

It would require to split whole method to multiple fragments.

So I did it this way:
- splitMeasure is acting always on masterScore (in local score there is wrapper to get local inserted measure)
- I removed option `addToAllScores` as it has to be always true now
- in only single case, where `addToAllScores = false` (only local score) were used: creating header frame for excerpts on scores without header frame on mastrer score I changed to be without `insertMeasure` method

@cbjeukendrup

File: src/engraving/dom/excerpt.cpp [hunk 1]
```diff
@@ -347,10 +347,12 @@ void Excerpt::createExcerpt(Excerpt* excerpt)
     MeasureBase* scoreMeasure = score->first();
 
     if (!scoreMeasure || !scoreMeasure->isVBox()) {
-        Score::InsertMeasureOptions options;
-        options.addToAllScores = false;
+        MeasureBase* newMeasureBase = toMeasureBase(Factory::createItem(ElementType::VBOX, score->dummy()));
```
Comment (cbjeukendrup) [inline src/engraving/dom/excerpt.cpp:350]:
```suggestion
        MeasureBase* newMeasureBase = Factory::createVBox(score->dummy()->system());
```

**Comment by cbjeukendrup** [comment]
@sammik That seems a good solution! But now a rebase is needed, to include the changes we made to `insertMeasure` in that other PR...

**Comment by sammik** [comment]
@cbjeukendrup done!

**Review by cbjeukendrup — DISMISSED** [review]

**Comment by sammik** [comment]
@cbjeukendrup 
it should be finished now

I am sorry, but I needed to change many things, so it requires another proper review

I split original `insertMeasure` into 
- `MasterScore::insertMeasure` which is used to inserting measures, and adds new measure to all scores
- `Score::insertBox` which is used to insert frames, and adds frame to single score only

File: src/engraving/dom/box.cpp [hunk 2]
```diff
@@ -431,7 +431,7 @@ void Box::manageExclusionFromParts(bool exclude)
                 continue;
             }
             MeasureBase* newMB = nullptr;
-            if (!excludeFromOtherParts()) {
+            if (!next()->isMeasure() && !next()->excludeFromOtherParts()) {
                 for (auto e : linkList()) {
```
Comment (cbjeukendrup) [inline src/engraving/dom/box.cpp:435]: Shouldn't this be `next()->linkList()` as well? (assuming that `newMB` is supposed to be the new `next()` for the new box that will be created)

Comment (sammik) [reply src/engraving/dom/box.cpp:435]: Yes, of course, thanks.

Comment (sammik) [reply src/engraving/dom/box.cpp:435]: @cbjeukendrup I was also thinking, if not to create new function `MeasureBase::getInScore(Score*, bool useNextMeasureFallback")` which returns MeasureBase of given score.

Than I could replace all that new code by `MeasureBase* newMB = next().getInScore(score, true);`

What do You think?

Comment (cbjeukendrup) [reply src/engraving/dom/box.cpp]: That sounds like a good idea!

File: src/engraving/dom/excerpt.cpp [hunk 3]
```diff
@@ -348,10 +348,7 @@ void Excerpt::createExcerpt(Excerpt* excerpt)
     MeasureBase* scoreMeasure = score->first();
 
     if ((!scoreMeasure || !scoreMeasure->isVBox()) && !masterMeasure->excludeFromOtherParts()) {
-        Score::InsertMeasureOptions options;
-        options.addToAllScores = false;
-
-        score->insertMeasure(ElementType::VBOX, scoreMeasure, options);
+        score->insertMeasure(ElementType::VBOX, scoreMeasure);
```
Comment (cbjeukendrup) [inline src/engraving/dom/excerpt.cpp:351]: I would prefer to use `insertBox` in those places where the type is explicitly a box type anyway

**Comment by cbjeukendrup** [comment]
Apart from these last two comments it looks good to me now!

**Comment by sammik** [comment]
> Apart from these last two comments it looks good to me now!

@cbjeukendrup resolved and rebased.

All changes are in new commits

**Review by cbjeukendrup — APPROVED** [review]

**Comment by oktophonie** [comment]
With this PR I can't get 'Insert [bar] before selection' to do anything at all.

**Comment by sammik** [comment]
> With this PR I can't get 'Insert [bar] before selection' to do anything at all.

Thx, Ill check it tonight.

**Comment by sammik** [comment]
@oktophonie done

**Review by oktophonie — APPROVED** [review]

**Review by cbjeukendrup — APPROVED** [review]

**Pull request merged**